### PR TITLE
fix: cron flow for canaries

### PR DESCRIPTION
**Pull request by yashmehrotra**
Fixes: #1105

### CODE REVIEW

File: pkg/jobs/canary/canary_jobs.go [hunk 1]
```diff
@@ -253,15 +256,26 @@ func SyncCanaryJob(canary v1.Canary) error {
...
+		logger.Infof("Scheduled %s: %s", canary, canary.Spec.GetSchedule())
+
+		canaryUpdateTimeCache[dbCanary.ID.String()] = dbCanary.UpdatedAt
 	}
-	logger.Infof("Scheduled %s: %s", canary, canary.Spec.GetSchedule())
 
-	entry = findCronEntry(canary)
-	if entry != nil && time.Until(entry.Next) < 1*time.Hour {
-		// run all regular canaries on startup
+	// Run canary on startup
```
Comment (moshloop) [inline pkg/jobs/canary/canary_jobs.go:277]: Run all regularly scheduled canaries on startup (<1h) and not daily/weekly schedules

File: pkg/jobs/canary/canary_jobs.go [hunk 2]
```diff
@@ -269,15 +272,26 @@ func SyncCanaryJob(canary v1.Canary) error {
...
+
+		canaryUpdateTimeCache[dbCanary.ID.String()] = dbCanary.UpdatedAt
 	}
-	logger.Infof("Scheduled %s: %s", canary, canary.Spec.GetSchedule())
 
-	entry = findCronEntry(canary)
-	if entry != nil && time.Until(entry.Next) < 1*time.Hour {
-		// run all regular canaries on startup
+	//Run all regularly scheduled canaries on startup (<1h) and not daily/weekly schedules
+	if entry != nil && time.Until(entry.Next) < 1*time.Hour && !exists {
```
Comment (adityathebe) [inline pkg/jobs/canary/canary_jobs.go:294]: Is this condition ever satisfied? If the cron entry is not nil, then the update time will also exist.

File: pkg/jobs/canary/canary_jobs.go [hunk 3]
```diff
@@ -269,15 +272,26 @@ func SyncCanaryJob(canary v1.Canary) error {
...
 		job = entry.Job.(CanaryJob)
```
Comment (adityathebe) [inline pkg/jobs/canary/canary_jobs.go:295]: Do we need this line?

File: pkg/jobs/canary/canary_jobs.go (hunk 2 above)
Comment (yashmehrotra) [reply pkg/jobs/canary/canary_jobs.go:295]: have updated the code to modify entry

File: pkg/jobs/canary/canary_jobs.go (hunk 3 above)
Comment (moshloop) [reply pkg/jobs/canary/canary_jobs.go]: Yes, from a ux perspective if I add a check that runs every 15 minutes I don't want to wait for 15m to see if it worked

Comment (adityathebe) [reply pkg/jobs/canary/canary_jobs.go]: @moshloop I meant just this one line

`job = entry.Job.(CanaryJob)`

File: pkg/jobs/canary/canary_jobs.go [hunk 4]
```diff
@@ -269,17 +272,28 @@ func SyncCanaryJob(canary v1.Canary) error {
...
+		if entry != nil {
+			CanaryScheduler.Remove(entry.ID)
+		}
+
+		// Schedule canary for the first time
+		entryID, err := CanaryScheduler.AddJob(canary.Spec.GetSchedule(), job)
+		if err != nil {
+			return fmt.Errorf("failed to schedule canary %s/%s: %v", canary.Namespace, canary.Name, err)
+		}
+		*entry = CanaryScheduler.Entry(entryID)
```
Comment (adityathebe) [inline pkg/jobs/canary/canary_jobs.go:288]: This panics in the first run when the `entry` var is nil.

Comment (moshloop) [reply pkg/jobs/canary/canary_jobs.go:288]: @yashmehrotra ^

**Pull request merged**